### PR TITLE
Add interpreter support for EH

### PR DESCRIPTION
**Pull request by aheejin**
This adds interpreter support for EH instructions. This adds
`ExceptionPackage` struct, which contains info of a thrown exception (an
event tag and thrown values), and the union in `Literal` can take a
`unique_ptr` to `ExceptionPackage`. We need a destructor, a copy
constructor, and an assignment operator for `Literal`, because the union
in `Literal` now has a member that cannot be trivially copied or
deleted.

### CODE REVIEW

File: src/wasm/literal.cpp [hunk 1]
```diff
@@ -33,6 +33,67 @@ Literal::Literal(const uint8_t init[16]) : type(Type::v128) {
...
+      i64 = other.i64;
+      return;
+    case Type::v128:
+      memcpy(&v128, other.v128, 16);
+      return;
+    case Type::funcref:
+      func = other.func;
+      return;
+    case Type::exnref:
+      new (&exn) auto(std::make_unique<ExceptionPackage>(*other.exn));
```
Comment (tlively) [inline src/wasm/literal.cpp:54]: This seems complicated. Does `exn = std::make_unique<ExceptionPackage>(*other.exn)` not work?

File: src/wasm/literal.cpp [hunk 2]
```diff
@@ -33,6 +33,67 @@ Literal::Literal(const uint8_t init[16]) : type(Type::v128) {
...
+    case Type::nullref:
+      return;
+    case Type::anyref:
+    case Type::unreachable:
+      WASM_UNREACHABLE("unexpected type");
+  }
+  WASM_UNREACHABLE("unexpected type");
+}
+
+Literal& Literal::operator=(const Literal& other) {
```
Comment (tlively) [inline src/wasm/literal.cpp:66]: Can one of these be rewritten in terms of the other to reduce duplication? I think the constructor could be `*this = other`, but I could be wrong.

File: test/spec/exception-handling.wast [hunk 3]
```diff
@@ -2,38 +2,110 @@
...
+    (throw $e1 (i32.const 3) (f32.const 3.5))
+  )
+
+  (func (export "rethrow_nullref")
     (rethrow (ref.null))
   )
 
-  (func (export "try_test") (result i32)
+  (func (export "try_nothrow") (result i32)
     (try (result i32)
```
Comment (tlively) [inline test/spec/exception-handling.wast:18]: Not for this PR, but I wonder if it would make sense to have syntax like 

```
(try
  (then ...)
  (catch ...)
)
```

To match the if-then-else syntax.

File: test/spec/exception-handling.wast [hunk 4]
```diff
@@ -2,38 +2,110 @@
...
+        (block $l0 (result i32)
+          (rethrow
+            (br_on_exn $l0 $e0 (local.get $exn))
+          )
+        )
+      )
+    )
+  )
+
+  (func (export "br_on_exn_match_multiple_values") (result i32 f32)
```
Comment (tlively) [inline test/spec/exception-handling.wast:69]: Should this also work with zero values?

File: test/spec/exception-handling.wast [hunk 5]
```diff
@@ -2,38 +2,110 @@
...
+(assert_trap (invoke "throw_single_value"))
+(assert_trap (invoke "throw_multiple_values"))
+(assert_trap (invoke "rethrow_nullref"))
+(assert_return (invoke "try_nothrow") (i32.const 3))
+(assert_return (invoke "try_throw_catch") (i32.const 3))
+(assert_trap (invoke "try_throw_rethrow"))
+(assert_trap (invoke "br_on_exn_nullref"))
+(assert_return (invoke "br_on_exn_match_single_value") (i32.const 5))
+(assert_return (invoke "br_on_exn_match_multiple_values") (tuple.make (i32.const 3) (f32.const 3.5)))
+(assert_trap (invoke "br_on_exn_dont_match"))
```
Comment (tlively) [inline test/spec/exception-handling.wast:108]: Another test to add would be throwing and catching across a function boundary, if that works.

File: src/literal.h [hunk 6]
```diff
@@ -128,6 +142,7 @@ class Literal {
   }
   std::array<uint8_t, 16> getv128() const;
   Name getFunc() const { return func; }
+  const ExceptionPackage& getExceptionPackage() const { return *exn.get(); }
```
Comment (kripken) [inline src/literal.h:145]: Can we check the `type` is correct in this getter? Seems like we could in the one before it too, so maybe I'm forgetting something...

Comment (aheejin) [reply src/literal.h:145]: Added assertions in both functions.  I added `getFunc()` when I added support for the reference types and I had forgotten to add an assertion then I think.

File: test/spec/exception-handling.wast (hunk 3 above)
Comment (aheejin) [reply test/spec/exception-handling.wast:18]: Hmm, not sure, because the try body is always executed, so it isn't something that depends on some condition, like `then` clause in `if`.

File: src/wasm/literal.cpp (hunk 2 above)
Comment (aheejin) [reply src/wasm/literal.cpp:66]: I think that works in this case. Done.

File: src/wasm/literal.cpp (hunk 1 above)
Comment (aheejin) [reply src/wasm/literal.cpp:54]: I mindlessly copied this from the implementation of the copy constructor, in which I tried to use placement new.. but anyway now we are calling `operator=` and this isn't even necessary. Done.

File: test/spec/exception-handling.wast (hunk 3 above)
Comment (binji) [reply test/spec/exception-handling.wast:18]: Rossberg was arguing for this in https://github.com/WebAssembly/exception-handling/issues/52 too. I think it makes sense, since the `catch` isn't actually nested within the try block.

File: src/wasm/literal.cpp (hunk 1 above)
Comment (aheejin) [reply src/wasm/literal.cpp:54]: No we can't do that, because now we are using this code in the copy constructor too. If we do `exn = ...` this will trigger the deletion of the current `exn`, which will in turn trigger deletion of its contained `ExceptionPackage`, which does not yet exist.

File: test/spec/exception-handling.wast (hunk 4 above)
Comment (aheejin) [reply test/spec/exception-handling.wast:69]: Yes. Added `br_on_exn_match_no_values` test for that.

File: test/spec/exception-handling.wast (hunk 5 above)
Comment (aheejin) [reply test/spec/exception-handling.wast:108]: Sounds good. Add several more tests to test this (throw in A and catch in B, throw in A in rethrow in B and catch in C, ...).

File: test/spec/exception-handling.wast (hunk 3 above)
Comment (aheejin) [reply test/spec/exception-handling.wast:19]: You mean, making it mandatory or optional? `if`s `then` and `else` are not mandatory.

**Comment by aheejin** [comment]
Sorry for the long silence. I've addressed the comments; PTAL.

File: test/spec/exception-handling.wast (hunk 3 above)
Comment (aheejin) [reply test/spec/exception-handling.wast:19]: If optional, `do` might make sense, as Rossberg suggested.

Comment (binji) [reply test/spec/exception-handling.wast:19]: I think we should make it mandatory. `then` and `else` used to be optional (and I think still are in wabt), but the spec requires them now (see https://webassembly.github.io/spec/core/text/instructions.html#folded-instructions)

Comment (aheejin) [reply test/spec/exception-handling.wast:19]: I see. Yeah, it makes sense. I reopened WebAssembly/exception-handling#52. I think this PR can land without that and we can make this change separately, if we decide to do that.
